OIDC authentication

jackmcguire1 opened the pull request:
Introduce Twitch OIDC authentication to Helix.

- Twitch user claims struct
- IDToken struct
- OIDC auth struct
- Request user OIDC access token
- UserInfo endpoint from access token

jackmcguire1 committed:
module name

- update module name

jackmcguire1 committed:
PROPOSAL: helix extension endpoints

- JWT signing
- extension configuration segments endpoints
- extension secrets endpoints
- UPDATE extension docs

jackmcguire1 committed:
HOTFIX: broadcaster ID in extension configuration requests

The Twitch docs are not really clear on some parts of the requests

getExtensionConfiguration:
 - broadcasterID is an optional query parameter

setExtensionConfiguration:
- broadcasterId is an optional body parameter
- validate the segment type if broadcasterID is provided

MISC:
- update 'GetExtensionSecret' -> 'GetExtensionSecrets'

jackmcguire1 committed:
OIDC authentication

- Twitch user claims struct
- IDToken struct
- OIDC auth struct
- Request user OIDC access token
- UserInfo endpoint from access token

nicklaw5 commented on go.mod:
```diff
@@ -1,3 +1,10 @@
 go 1.15
+
+require (
+	github.com/coreos/go-oidc v2.2.1+incompatible // indirect
+	github.com/coreos/go-oidc/v3 v3.0.0 // indirect
+	github.com/pquerna/cachecontrol v0.1.0 // indirect
+	golang.org/x/oauth2 v0.0.0-20210819190943-2bc19b11175f // indirect
```
This doesn't look right. Can you please run a `go mod tidy` to fix this.

nicklaw5 commented on authentication.go:
```diff
@@ -223,3 +254,89 @@ func (c *Client) ValidateToken(accessToken string) (bool, *ValidateTokenResponse
+	return
+}
+
+func (c *Client) UserInfoFromAccessToken(
```
```suggestion
func (c *Client) UserInfoFromOIDCAccessToken(
```

nicklaw5 reviewed (changes requested):
We need some documentation and some mock tests if it's not too much trouble.

jackmcguire1 committed:
HOTFIX: Create Event Subs - Total Field

The response contains a json field 'total' which we're
not currently unmarshalling to create event subs

jackmcguire1 committed:
Extension Live Channels

- Extension Live Channels
- Unit test for Extension Live Channels
- Supported Endpoints

jackmcguire1 committed:
feat: Send Chat Announcements

https://dev.twitch.tv/docs/api/reference#send-chat-announcement

- add api parameters, response, func to chat.go
- add unit tests

jackmcguire1 committed:
feat: Send Chat Announcements

https://dev.twitch.tv/docs/api/reference#send-chat-announcement

- add api parameters, response, func to chat.go
- add unit tests

jackmcguire1 committed:
Merge branch 'send-chat-announcement' into main

jackmcguire1 committed:
Auto stash before rebase of "nick"

jackmcguire1 committed:
remove main.go

jackmcguire1 committed:
Merge branch 'send-chat-announcement' into main

jackmcguire1 committed:
Merge branch 'charity' into main

jackmcguire1 committed:
fix: duplicate funcs and structs

- fixs

jackmcguire1 committed:
Fix: charity campaigns uri

- add prefix of '/' to the charity campaigns url

jackmcguire1 committed:
Merge branch 'fix' into main

jackmcguire1 committed:
fix: improvements

- oidc provider as interface
- refactor auth url to default url
- fetch auth url and oidc provider from opts of client
